bftengine/src/bftengine/ReplicaFactory.cpp

```diff
@@ -0,0 +1,289 @@
+// Concord
+//
+// Copyright (c) 2018-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "ReplicaFactory.hpp"
+#include "DebugPersistentStorage.hpp"
+#include "DbCheckpointManager.hpp"
+#include "IncomingMsgsStorageImp.hpp"
+#include "MsgReceiver.hpp"
+#include "PreProcessor.hpp"
+
+bftEngine::IReservedPages *bftEngine::ReservedPagesClientBase::res_pages_ = nullptr;
+namespace bftEngine {
+
+namespace impl {
+
+class ReplicaInternal : public IReplica {
+  friend class IReplica;
+
+ public:
+  ReplicaInternal(std::unique_ptr<preprocessor::PreProcessor> &&preprocessor = nullptr,
+                  const std::shared_ptr<concord::cron::TicksGenerator> &ticks_gen = nullptr,
+                  const std::shared_ptr<PersistentStorage> &persistent_storage = nullptr,
+                  const std::shared_ptr<IInternalBFTClient> &internal_client = nullptr,
+                  const std::function<void(bool)> &viewChangeCallBack = nullptr)
+      : preprocessor_{std::move(preprocessor)},
+        ticks_gen_{ticks_gen},
+        persistent_storage_{persistent_storage},
+        internal_client_{internal_client},
+        viewChangeCallBack_(viewChangeCallBack) {}
+  bool isRunning() const override;
+  int64_t getLastExecutedSequenceNum() const override { return replica_->getLastExecutedSequenceNum(); }
+  void start() override;
+  void stop() override;
+  void SetAggregator(std::shared_ptr<concordMetrics::Aggregator> a) override;
+  void restartForDebug(uint32_t delayMillis) override;
+  std::shared_ptr<concord::cron::TicksGenerator> ticksGenerator() const override;
+  std::shared_ptr<PersistentStorage> persistentStorage() const override;
+  std::shared_ptr<IInternalBFTClient> internalClient() const override { return internal_client_; }
+  void setReplica(std::unique_ptr<ReplicaBase> &&replica) { replica_ = std::move(replica); }
+  ReplicaBase *getReplica() { return replica_.get(); }
+
+ private:
+  std::unique_ptr<preprocessor::PreProcessor> preprocessor_;
```

Comment:
The PreProcessor class should not be a part of the replica and the replica should not know about it. It could probably be injected into the replica. Sorry, I missed this fact before.

Reply:
@glevkovich, that AFAIR that was our temporary idea. Correct?

Reply:
Besides keeping it in the factory (as on one point it was implemented) i don't see other way to this.

Reply:
@glevkovich That's exactly the point. Not ReplicaFactory keeps the PreProcessor object, but ReplicaInternal.

Reply:
Done